Respect thread_count argument in multipart uploads

kspector opened the pull request:
## Observed Behavior ##
A method call in form:

`Aws::S3::Object.new(...).upload(some_file, thread_count: 20)`

only utilizes 10 threads to perform the upload.

## Reasoning ##
[MultipartFileUploader](https://github.com/aws/aws-sdk-ruby/blob/version-3/gems/aws-sdk-s3/lib/aws-sdk-s3/multipart_file_uploader.rb#L37) accepts a `thread_count` option in its initializer, which sets instance variable `@thread_count` - however, [FileUploader](https://github.com/aws/aws-sdk-ruby/blob/version-3/gems/aws-sdk-s3/lib/aws-sdk-s3/file_uploader.rb#L42) only passes its own `@options` to the `MultipartFileUploader` initializer, which does not include a`thread_count` key value pair.

The `thread_count` option is therefore ignored due to the preference in method [upload_in_threads](https://github.com/aws/aws-sdk-ruby/blob/version-3/gems/aws-sdk-s3/lib/aws-sdk-s3/multipart_file_uploader.rb#L150) for the instance variable value of `@thread_count`.

## Solution ##
I moved the setting of instance variable `@thread_count` to the [upload](https://github.com/kspector/aws-sdk-ruby/blob/a2bf07e28fddaa8534e6f24afd8253651f5e09f6/gems/aws-sdk-s3/lib/aws-sdk-s3/multipart_file_uploader.rb#L54) method, which mirrors the location of its counterpart in [FileDownloader -> download](https://github.com/aws/aws-sdk-ruby/blob/version-3/gems/aws-sdk-s3/lib/aws-sdk-s3/file_downloader.rb#L27C9-L27C63).

I also added two tests to assert the default and specific cases, the [latter of which](https://github.com/kspector/aws-sdk-ruby/blob/a2bf07e28fddaa8534e6f24afd8253651f5e09f6/gems/aws-sdk-s3/spec/object/upload_file_spec.rb#L195-L201) fails against the current state of the codebase.

kspector committed:
Respect thread count argument for multipart upload

This option is otherwise ignored due to the preference in method upload_in_threads for the instance variable value of @thread_count.

alextwoods commented on gems/aws-sdk-s3/CHANGELOG.md:
```diff
@@ -1,6 +1,8 @@
 Unreleased Changes
 ------------------
 
+* Feature - Fix bug where thread_count option was not being respected for multipart uploads.
```
I'd classify this as `Issue` instead of feature since its correcting a bug.

alextwoods commented on gems/aws-sdk-s3/lib/aws-sdk-s3/multipart_file_uploader.rb:
```diff
@@ -31,10 +31,8 @@ class MultipartFileUploader
-      # @option options [Integer] :thread_count (THREAD_COUNT)
       def initialize(options = {})
         @client = options[:client] || Client.new
-        @thread_count = options[:thread_count] || THREAD_COUNT
```
I think we still want to set `@thread_count` here - otherwise this could be a breaking change.  However - I think we can skip setting a default ie, just `@thread_count = options[:thread_count]`.

Then in the `upload_in_threads` method we can do something like:
```ruby
      def upload_in_threads(pending, completed, options)
        # code skipped for brevity here
     
        (@thread_count || options[:thread_count] || THREAD_COUNT).times do
          thread = Thread.new do
```

kspector replied:
makes sense 👍

mullermp replied:
This works, but I would have suggested the opposite? The instance variable for the class is still valid, either as a passed value or a default. Any request level options should just supersede it. I would have recommended that upload_in_threads did something like `options.fetch(:thread_count, @thread_count)` and still have `@thread_count` as a user passed class option or a default to `THREAD_COUNT`. This way the hierarchy is preserved.

alextwoods replied:
Edit: Sorry, for the initial suggestion - I do agree with @mullermp's suggestion.  We can keep the defaulting of `@thread_count` in the initializer and use an `options.fetch(:thread_count, @thread_count)` in the `upload_in_threads` method.  I agree that preserves the normal order of precedence better.

kspector replied:
👍 Addressed in my latest commit @mullermp @alextwoods.

For consistency, I elected to implement the same variable precedence in `MultipartStreamUploader` as well.

alextwoods reviewed (commented):
Thanks for submitting this PR!  

Left a couple of comments.  I think we should also make the same change in the `MultipartStreamUploader` as well.

kspector committed:
Address PR feedback

kspector committed:
Add thread_count spec for upload_stream

kspector commented:
Thanks for the quick review @alextwoods, I addressed your feedback. Just let me know if the commit history needs to be cleaned too.

> I think we should also make the same change in the MultipartStreamUploader as well.

It seems that the only initialization of a [`MultipartStreamUploader` in the repo](https://github.com/aws/aws-sdk-ruby/blob/d1565421f4cc4321e35f57eba1cab523cf4fe0e9/gems/aws-sdk-s3/lib/aws-sdk-s3/customizations/object.rb#L387C20-L387C43) _does_ pass the `thread_count` option, so I held off on making the equivalent change there in favor of [an additional spec](https://github.com/aws/aws-sdk-ruby/pull/3000/commits/07ef78cfe975737f3da2b521573f644487649665).

kspector committed:
Prefer request-level thread_count, when available

alextwoods reviewed (approved):
Nice - this looks good!  Thanks for submitting and making updates!

alextwoods commented:
I've merged, this should be released today.
